File: tools/agent-trace/go/stats.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"text/tabwriter"
)

type TokenTotals struct {
	Input         int64 `json:"input"`
	Output        int64 `json:"output"`
	CacheRead     int64 `json:"cacheRead"`
	CacheCreation int64 `json:"cacheCreation"`
}

type StatsAggregate struct {
	Root             string                   `json:"root"`
	Files            int                      `json:"files"`
	Sessions         int                      `json:"sessions"`
	Records          int                      `json:"records"`
	Malformed        int                      `json:"malformed"`
	RecordsByType    map[string]int           `json:"recordsByType"`
	ToolUses         map[string]int           `json:"toolUses"`
	Models           map[string]int           `json:"models"`
	Tokens           TokenTotals              `json:"tokens"`
	APIErrors        int                      `json:"apiErrors"`
	ToolResultErrors int                      `json:"toolResultErrors"`
	Projects         map[string]*ProjectStats `json:"projects"`
	sessions         map[string]struct{}      `json:"-"`
}

type ProjectStats struct {
	Files            int            `json:"files"`
	Sessions         int            `json:"sessions"`
	Records          int            `json:"records"`
	Malformed        int            `json:"malformed"`
	RecordsByType    map[string]int `json:"recordsByType"`
	APIErrors        int            `json:"apiErrors"`
	ToolResultErrors int            `json:"toolResultErrors"`
	Tokens           TokenTotals    `json:"tokens"`
	sessions         map[string]struct{}
}

func NewStats(root string) *StatsAggregate {
	return &StatsAggregate{
		Root:          root,
		RecordsByType: make(map[string]int),
		ToolUses:      make(map[string]int),
		Models:        make(map[string]int),
		Projects:      make(map[string]*ProjectStats),
		sessions:      make(map[string]struct{}),
	}
}

func (s *StatsAggregate) project(name string) *ProjectStats {
	project := s.Projects[name]
	if project == nil {
		project = &ProjectStats{RecordsByType: make(map[string]int), sessions: make(map[string]struct{})}
		s.Projects[name] = project
	}
	return project
}

func RunStats(root, format string) error {
	stats := NewStats(root)
	err := WalkTranscripts(root, func(path, project string) error {
		stats.Files++
		projectStats := stats.project(project)
		projectStats.Files++
		fileStats, err := ParseFile(path, project, func(record *Record, ctx ParseContext) error {
			stats.AddRecord(record, ctx)
			return nil
		})
		stats.Malformed += fileStats.Malformed
		projectStats.Malformed += fileStats.Malformed
		return err
	})
	if err != nil {
		return err
	}
	stats.Finish()
	if format == "json" {
		encoder := json.NewEncoder(os.Stdout)
		encoder.SetIndent("", "  ")
		return encoder.Encode(stats)
	}
	return stats.PrintTable()
}

func (s *StatsAggregate) AddRecord(record *Record, ctx ParseContext) {
	projectStats := s.project(ctx.Project)
	s.Records++
	projectStats.Records++

	sessionID := sessionIDFor(record, ctx.Path)
	s.sessions[sessionID] = struct{}{}
	projectStats.sessions[sessionID] = struct{}{}

	recordType := bucketRecordType(record.Type)
	s.RecordsByType[recordType]++
	projectStats.RecordsByType[recordType]++

	if record.Type == "assistant" && record.Message != nil {
		if record.Message.Model != "" {
			s.Models[record.Message.Model]++
		}
		addUsage(&s.Tokens, record.Message.Usage)
		addUsage(&projectStats.Tokens, record.Message.Usage)
	}
	for _, name := range record.ToolUseNames() {
		s.ToolUses[name]++
	}
	if record.HasAPIError() {
		s.APIErrors++
		projectStats.APIErrors++
	}
	if record.HasToolError() {
		s.ToolResultErrors++
		projectStats.ToolResultErrors++
	}
}

func (s *StatsAggregate) Finish() {
	s.Sessions = len(s.sessions)
	for _, project := range s.Projects {
		project.Sessions = len(project.sessions)
	}
}

func addUsage(tokens *TokenTotals, usage *Usage) {
	if usage == nil {
		return
	}
	if usage.InputTokens != nil {
		tokens.Input += *usage.InputTokens
	}
	if usage.OutputTokens != nil {
		tokens.Output += *usage.OutputTokens
	}
	if usage.CacheReadInputTokens != nil {
		tokens.CacheRead += *usage.CacheReadInputTokens
	}
	if usage.CacheCreationInputTokens != nil {
		tokens.CacheCreation += *usage.CacheCreationInputTokens
	}
}

func (s *StatsAggregate) PrintTable() error {
	writer := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
	fmt.Fprintf(writer, "Root:\t%s\n", s.Root)
	fmt.Fprintf(writer, "Files:\t%d\n", s.Files)
	fmt.Fprintf(writer, "Sessions:\t%d\n", s.Sessions)
	fmt.Fprintf(writer, "Records:\t%d\n", s.Records)
	fmt.Fprintf(writer, "Malformed:\t%d\n", s.Malformed)
	fmt.Fprintf(writer, "API errors:\t%d\n", s.APIErrors)
	fmt.Fprintf(writer, "Tool result errors:\t%d\n", s.ToolResultErrors)
	fmt.Fprintf(writer, "Tokens:\tinput=%d output=%d cache-read=%d cache-creation=%d\n", s.Tokens.Input, s.Tokens.Output, s.Tokens.CacheRead, s.Tokens.CacheCreation)

	fmt.Fprintln(writer, "\nRecords by type")
	fmt.Fprintln(writer, "TYPE\tCOUNT")
	for _, item := range sortedCounts(s.RecordsByType) {
		fmt.Fprintf(writer, "%s\t%d\n", item.Key, item.Value)
	}

	fmt.Fprintln(writer, "\nTool use histogram")
	fmt.Fprintln(writer, "TOOL\tCOUNT")
	for _, item := range sortedCounts(s.ToolUses) {
		fmt.Fprintf(writer, "%s\t%d\n", item.Key, item.Value)
	}

	fmt.Fprintln(writer, "\nModels")
	fmt.Fprintln(writer, "MODEL\tCOUNT")
	for _, item := range sortedCounts(s.Models) {
		fmt.Fprintf(writer, "%s\t%d\n", item.Key, item.Value)
	}

	fmt.Fprintln(writer, "\nProjects")
	fmt.Fprintln(writer, "PROJECT\tFILES\tSESSIONS\tRECORDS\tMALFORMED\tAPI_ERRORS\tTOOL_ERRORS\tINPUT\tOUTPUT")
	projectNames := make([]string, 0, len(s.Projects))
	for name := range s.Projects {
		projectNames = append(projectNames, name)
	}
	sort.Strings(projectNames)
	for _, name := range projectNames {
		project := s.Projects[name]
		fmt.Fprintf(writer, "%s\t%d\t%d\t%d\t%d\t%d\t%d\t%d\t%d\n", name, project.Files, project.Sessions, project.Records, project.Malformed, project.APIErrors, project.ToolResultErrors, project.Tokens.Input, project.Tokens.Output)
	}
	return writer.Flush()
}

type countItem struct {
	Key   string
	Value int
}

func sortedCounts(counts map[string]int) []countItem {
	items := make([]countItem, 0, len(counts))
	for key, value := range counts {
		items = append(items, countItem{Key: key, Value: value})
	}
	sort.Slice(items, func(i, j int) bool {
		if items[i].Value == items[j].Value {
			return items[i].Key < items[j].Key
		}
		return items[i].Value > items[j].Value
	})
	return items
}

File: tools/agent-trace/rust/src/stats.rs
use std::collections::{BTreeMap, BTreeSet};
use std::path::Path;

use anyhow::Result;
use serde::Serialize;

use crate::model::{Record, Usage};
use crate::parse::{self, TranscriptFile};

#[derive(Debug, Clone, Default, Serialize)]
pub struct TokenTotals {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_creation_input_tokens: u64,
    pub cache_read_input_tokens: u64,
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct ProjectStats {
    pub files: usize,
    pub sessions: usize,
    pub records: u64,
    pub malformed: u64,
    pub api_errors: u64,
    pub tool_result_errors: u64,
    pub tokens: TokenTotals,
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct StatsReport {
    pub root: String,
    pub files: usize,
    pub sessions: usize,
    pub records: u64,
    pub malformed: u64,
    pub records_by_type: BTreeMap<String, u64>,
    pub tool_uses: BTreeMap<String, u64>,
    pub models: BTreeMap<String, u64>,
    pub tokens: TokenTotals,
    pub api_errors: u64,
    pub tool_result_errors: u64,
    pub projects: BTreeMap<String, ProjectStats>,
}

#[derive(Default)]
struct StatsBuilder {
    report: StatsReport,
    sessions: BTreeSet<String>,
    project_sessions: BTreeMap<String, BTreeSet<String>>,
}

pub fn collect(root: &Path) -> Result<StatsReport> {
    let transcripts = parse::discover_transcripts(root)?;
    let mut builder = StatsBuilder::new(root, transcripts.len());

    for transcript in transcripts {
        builder.add_file(&transcript);
        let fallback_session = parse::session_fallback(&transcript.path);
        let project = transcript.project.clone();
        let report = parse::parse_jsonl_file(&transcript.path, |parsed| {
            let session_id = parsed
                .record
                .session_id
                .clone()
                .unwrap_or_else(|| fallback_session.clone());
            builder.add_record(&project, &session_id, &parsed.record);
            Ok(())
        })?;
        builder.add_malformed(&project, report.malformed as u64);
    }

    Ok(builder.finish())
}

impl StatsBuilder {
    fn new(root: &Path, file_count: usize) -> Self {
        let mut report = StatsReport {
            root: root.display().to_string(),
            files: file_count,
            ..StatsReport::default()
        };
        for record_type in [
            "assistant",
            "user",
            "attachment",
            "last-prompt",
            "queue-operation",
            "mode",
            "system",
            "pr-link",
            "custom-title",
            "summary",
            "other",
        ] {
            report.records_by_type.insert(record_type.to_string(), 0);
        }

        Self {
            report,
            sessions: BTreeSet::new(),
            project_sessions: BTreeMap::new(),
        }
    }

    fn add_file(&mut self, transcript: &TranscriptFile) {
        self.report
            .projects
            .entry(transcript.project.clone())
            .or_default()
            .files += 1;
    }

    fn add_record(&mut self, project: &str, session_id: &str, record: &Record) {
        self.sessions.insert(session_id.to_string());
        self.project_sessions
            .entry(project.to_string())
            .or_default()
            .insert(session_id.to_string());

        self.report.records += 1;
        *self
            .report
            .records_by_type
            .entry(record.type_bucket().to_string())
            .or_default() += 1;

        let project_stats = self.report.projects.entry(project.to_string()).or_default();
        project_stats.records += 1;

        if record.is_api_error() {
            self.report.api_errors += 1;
            project_stats.api_errors += 1;
        }

        let tool_errors = record.tool_result_error_count() as u64;
        if tool_errors > 0 {
            self.report.tool_result_errors += tool_errors;
            project_stats.tool_result_errors += tool_errors;
        }

        if let Some(message) = record.message.as_ref() {
            if let Some(model) = message.model.as_deref() {
                *self.report.models.entry(model.to_string()).or_default() += 1;
            }
            if let Some(usage) = message.usage.as_ref() {
                add_usage(&mut self.report.tokens, usage);
                add_usage(&mut project_stats.tokens, usage);
            }
        }

        for tool in record.tool_names() {
            *self.report.tool_uses.entry(tool.to_string()).or_default() += 1;
        }
    }

    fn add_malformed(&mut self, project: &str, malformed: u64) {
        self.report.malformed += malformed;
        self.report
            .projects
            .entry(project.to_string())
            .or_default()
            .malformed += malformed;
    }

    fn finish(mut self) -> StatsReport {
        self.report.sessions = self.sessions.len();
        for (project, sessions) in self.project_sessions {
            self.report.projects.entry(project).or_default().sessions = sessions.len();
        }
        self.report
    }
}

fn add_usage(total: &mut TokenTotals, usage: &Usage) {
    total.input_tokens += usage.input_tokens.unwrap_or(0);
    total.output_tokens += usage.output_tokens.unwrap_or(0);
    total.cache_creation_input_tokens += usage.cache_creation_input_tokens.unwrap_or(0);
    total.cache_read_input_tokens += usage.cache_read_input_tokens.unwrap_or(0);
}

pub fn render_table(report: &StatsReport) -> String {
    let mut out = String::new();
    out.push_str("Agent Trace Stats\n");
    out.push_str(&format!("root: {}\n", report.root));
    out.push_str(&format!("files: {}\n", report.files));
    out.push_str(&format!("sessions: {}\n", report.sessions));
    out.push_str(&format!("records: {}\n", report.records));
    out.push_str(&format!("malformed: {}\n", report.malformed));
    out.push_str(&format!("api_errors: {}\n", report.api_errors));
    out.push_str(&format!(
        "tool_result_errors: {}\n\n",
        report.tool_result_errors
    ));

    push_map_section(
        &mut out,
        "Records by type",
        "type",
        "count",
        &report.records_by_type,
    );
    push_map_section(&mut out, "Models", "model", "count", &report.models);
    push_map_section(
        &mut out,
        "Tool use histogram",
        "tool",
        "count",
        &report.tool_uses,
    );

    out.push_str("Token totals\n");
    out.push_str(&format!(
        "  {:<30} {:>15}\n",
        "input_tokens", report.tokens.input_tokens
    ));
    out.push_str(&format!(
        "  {:<30} {:>15}\n",
        "output_tokens", report.tokens.output_tokens
    ));
    out.push_str(&format!(
        "  {:<30} {:>15}\n",
        "cache_creation_input_tokens", report.tokens.cache_creation_input_tokens
    ));
    out.push_str(&format!(
        "  {:<30} {:>15}\n\n",
        "cache_read_input_tokens", report.tokens.cache_read_input_tokens
    ));

    out.push_str("Projects\n");
    out.push_str(&format!(
        "  {:<55} {:>7} {:>8} {:>10} {:>10} {:>10} {:>10} {:>14} {:>14}\n",
        "project",
        "files",
        "sessions",
        "records",
        "malformed",
        "api_err",
        "tool_err",
        "input_tok",
        "output_tok"
    ));
    for (project, stats) in &report.projects {
        out.push_str(&format!(
            "  {:<55} {:>7} {:>8} {:>10} {:>10} {:>10} {:>10} {:>14} {:>14}\n",
            truncate_middle(project, 55),
            stats.files,
            stats.sessions,
            stats.records,
            stats.malformed,
            stats.api_errors,
            stats.tool_result_errors,
            stats.tokens.input_tokens,
            stats.tokens.output_tokens
        ));
    }

    out
}

fn push_map_section(
    out: &mut String,
    title: &str,
    key_header: &str,
    value_header: &str,
    map: &BTreeMap<String, u64>,
) {
    out.push_str(title);
    out.push('\n');
    out.push_str(&format!("  {:<32} {:>15}\n", key_header, value_header));
    for (key, value) in map {
        if *value == 0 && title != "Records by type" {
            continue;
        }
        out.push_str(&format!(
            "  {:<32} {:>15}\n",
            truncate_middle(key, 32),
            value
        ));
    }
    out.push('\n');
}

fn truncate_middle(value: &str, width: usize) -> String {
    let len = value.chars().count();
    if len <= width {
        return value.to_string();
    }
    if width <= 1 {
        return "…".to_string();
    }

    let left = (width - 1) / 2;
    let right = width - 1 - left;
    let prefix: String = value.chars().take(left).collect();
    let suffix: String = value
        .chars()
        .rev()
        .take(right)
        .collect::<Vec<_>>()
        .into_iter()
        .rev()
        .collect();
    format!("{prefix}…{suffix}")
}
